File: clients/ios/SsdidDrive/SsdidDrive/Domain/Model/Recovery.swift
import Foundation

/// Recovery configuration for a user
struct RecoveryConfig: Codable, Equatable {
    let isConfigured: Bool
    let threshold: Int
    let totalShares: Int
    let trustees: [Trustee]
    let createdAt: Date?

    enum CodingKeys: String, CodingKey {
        case isConfigured = "is_configured"
        case threshold
        case totalShares = "total_shares"
        case trustees
        case createdAt = "created_at"
    }

    /// Empty/not configured state
    static let notConfigured = RecoveryConfig(
        isConfigured: false,
        threshold: 0,
        totalShares: 0,
        trustees: [],
        createdAt: nil
    )
}

/// A trustee who holds a recovery share
struct Trustee: Codable, Identifiable, Equatable, Hashable {
    let id: String
    let userId: String
    let email: String
    let displayName: String?
    let hasAccepted: Bool
    let acceptedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case email
        case displayName = "display_name"
        case hasAccepted = "has_accepted"
        case acceptedAt = "accepted_at"
    }
}

/// A recovery share (Shamir secret sharing)
struct RecoveryShare: Codable, Identifiable, Equatable {
    let id: String
    let trusteeId: String
    let encryptedShare: Data
    let shareIndex: Int
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case trusteeId = "trustee_id"
        case encryptedShare = "encrypted_share"
        case shareIndex = "share_index"
        case createdAt = "created_at"
    }
}

/// A pending recovery request
struct RecoveryRequest: Codable, Identifiable, Equatable {
    let id: String
    let requesterId: String
    let requesterEmail: String
    let requesterName: String?
    let status: Status
    let approvedShares: Int
    let requiredShares: Int
    let expiresAt: Date
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case requesterId = "requester_id"
        case requesterEmail = "requester_email"
        case requesterName = "requester_name"
        case status
        case approvedShares = "approved_shares"
        case requiredShares = "required_shares"
        case expiresAt = "expires_at"
        case createdAt = "created_at"
    }

    enum Status: String, Codable {
        case pending
        case approved
        case rejected
        case expired
        case completed

        var displayName: String {
            rawValue.capitalized
        }
    }

    /// Progress towards threshold
    var progress: Double {
        guard requiredShares > 0 else { return 0 }
        return Double(approvedShares) / Double(requiredShares)
    }

    /// Has enough shares been collected
    var hasThresholdMet: Bool {
        approvedShares >= requiredShares
    }
}

/// Request to setup recovery
struct SetupRecoveryRequest: Codable {
    let threshold: Int
    let shares: [TrusteeShare]

    struct TrusteeShare: Codable {
        let trusteeEmail: String
        let encryptedShare: Data

        enum CodingKeys: String, CodingKey {
            case trusteeEmail = "trustee_email"
            case encryptedShare = "encrypted_share"
        }
    }
}

/// Response from recovery setup
struct SetupRecoveryResponse: Codable {
    let config: RecoveryConfig
}

/// Pending requests response
struct PendingRequestsResponse: Codable {
    let requests: [RecoveryRequest]
}

/// Request to approve recovery
struct ApproveRecoveryRequest: Codable {
    let share: Data
}

/// Response from initiating recovery
struct InitiateRecoveryResponse: Codable {
    let request: RecoveryRequest
}

/// Response when recovery is complete
struct RecoveryCompleteResponse: Codable {
    let masterKey: Data
    let message: String

    enum CodingKeys: String, CodingKey {
        case masterKey = "master_key"
        case message
    }
}

File: clients/android/app/src/main/kotlin/com/securesharing/domain/model/Recovery.kt
package com.securesharing.domain.model

import java.time.Instant

/**
 * Recovery configuration for a user's account.
 *
 * Defines the Shamir secret sharing parameters:
 * - threshold: Minimum shares required to recover
 * - totalShares: Total shares distributed to trustees
 */
data class RecoveryConfig(
    val id: String,
    val userId: String,
    val threshold: Int,
    val totalShares: Int,
    val status: RecoveryConfigStatus,
    val createdAt: Instant,
    val updatedAt: Instant
) {
    /**
     * Check if recovery is fully configured (all shares distributed).
     */
    fun isConfigured(): Boolean = status == RecoveryConfigStatus.ACTIVE

    /**
     * Check if more shares can be added.
     */
    fun canAddShares(): Boolean = status == RecoveryConfigStatus.PENDING
}

enum class RecoveryConfigStatus {
    PENDING,    // Shares not yet distributed
    ACTIVE,     // All shares distributed, recovery available
    DISABLED;   // Recovery disabled

    companion object {
        fun fromString(value: String): RecoveryConfigStatus {
            return when (value.lowercase()) {
                "pending" -> PENDING
                "active" -> ACTIVE
                "disabled" -> DISABLED
                else -> PENDING
            }
        }
    }
}

/**
 * A recovery share held by a trustee.
 *
 * The share is encrypted for the trustee and can only be decrypted
 * by them to approve a recovery request.
 */
data class RecoveryShare(
    val id: String,
    val configId: String,
    val grantorId: String,
    val trusteeId: String,
    val shareIndex: Int,
    val status: RecoveryShareStatus,
    val grantor: User?,
    val trustee: User?,
    val createdAt: Instant,
    val updatedAt: Instant
) {
    /**
     * Check if this share can be used for recovery.
     */
    fun isActive(): Boolean = status == RecoveryShareStatus.ACCEPTED
}

enum class RecoveryShareStatus {
    PENDING,    // Trustee hasn't accepted yet
    ACCEPTED,   // Trustee accepted, share is active
    REJECTED,   // Trustee rejected
    REVOKED;    // Owner revoked

    companion object {
        fun fromString(value: String): RecoveryShareStatus {
            return when (value.lowercase()) {
                "pending" -> PENDING
                "accepted" -> ACCEPTED
                "rejected" -> REJECTED
                "revoked" -> REVOKED
                else -> PENDING
            }
        }
    }
}

/**
 * A recovery request initiated when a user needs to recover their account.
 *
 * The request goes through a flow:
 * 1. User initiates with new public keys
 * 2. Trustees approve and submit re-encrypted shares
 * 3. Once threshold met, user can complete recovery
 */
data class RecoveryRequest(
    val id: String,
    val userId: String,
    val status: RecoveryRequestStatus,
    val reason: String?,
    val user: User?,
    val progress: RecoveryProgress?,
    val createdAt: Instant,
    val updatedAt: Instant
) {
    /**
     * Check if this request is waiting for approvals.
     */
    fun isPending(): Boolean = status == RecoveryRequestStatus.PENDING

    /**
     * Check if this request has enough approvals to complete.
     */
    fun canComplete(): Boolean = status == RecoveryRequestStatus.APPROVED

    /**
     * Check if this request has been completed.
     */
    fun isCompleted(): Boolean = status == RecoveryRequestStatus.COMPLETED
}

enum class RecoveryRequestStatus {
    PENDING,    // Waiting for trustee approvals
    APPROVED,   // Threshold met, ready to complete
    COMPLETED,  // Recovery completed
    REJECTED,   // Request rejected/expired
    CANCELLED;  // User cancelled

    companion object {
        fun fromString(value: String): RecoveryRequestStatus {
            return when (value.lowercase()) {
                "pending" -> PENDING
                "approved" -> APPROVED
                "completed" -> COMPLETED
                "rejected" -> REJECTED
                "cancelled" -> CANCELLED
                else -> PENDING
            }
        }
    }
}

/**
 * Progress tracking for a recovery request.
 */
data class RecoveryProgress(
    val threshold: Int,
    val approvals: Int,
    val remaining: Int
) {
    /**
     * Check if threshold has been met.
     */
    fun thresholdMet(): Boolean = approvals >= threshold

    /**
     * Get completion percentage.
     */
    fun percentage(): Float = if (threshold > 0) {
        (approvals.toFloat() / threshold * 100).coerceAtMost(100f)
    } else {
        0f
    }
}

/**
 * An approval for a recovery request from a trustee.
 */
data class RecoveryApproval(
    val id: String,
    val requestId: String,
    val shareId: String,
    val approverId: String,
    val createdAt: Instant
)
